hwilib/devices/tapsigner.py

```diff
@@ -0,0 +1,474 @@
+    """Do not allow hardened components after non-hardened"""
+    found_non_hardened = False
+    for component in path:
+        curr_hardened = is_hardened(component)
+        # have we found non hardened component?
+        curr_non_hardened = not curr_hardened
+        if curr_hardened is True and found_non_hardened is True:
+            raise ValueError(f"Hardened path component after non-hardened {path}")
+        found_non_hardened = curr_non_hardened
+
+
+def split_bip32_path(path: Sequence[int]) -> Tuple[List[int], List[int]]:
+    """Split bip32 path to hardened part and non-hardened part"""
+    hardened = []
+    non_hardened = []
+    for component in path:
+        if is_hardened(component):
+            hardened.append(component)
+        else:
+            non_hardened.append(component)
+    return hardened, non_hardened
+
+
+class TapsignerClient(HardwareWalletClient):
+    # In path variable we store card ident
+    # In password variable we store card CVC
+    def __init__(self, path: str, password: str = "", expert: bool = False, chain: Chain = Chain.MAIN) -> None:
+        super(TapsignerClient, self).__init__(path, password, expert, chain)
+        self.device = find_card_by_path(path=path)
+
+    def interactive_get_cvc(self) -> None:
+        if not self.password:
+            cvc = getpass.getpass("Please provide CVC:\n")
+            if cvc:
+                self.password = cvc
+
+    def get_master_xpub(self, addrtype: AddressType = AddressType.WIT, account: int = 0) -> ExtendedKey:
+        bip44_pupose = get_bip44_purpose(addrtype)
+        bip44_chain = get_bip44_chain(self.chain)
+        path = f"m/{bip44_pupose}h/{bip44_chain}h/{account}h"
+        return self.get_pubkey_at_path(path)
+
+    def get_master_fingerprint(self) -> bytes:
+        """
+        m fingerprint
+        """
+        extended_key_m = self.get_pubkey_at_path("m")
+        fingerprint = hash160(extended_key_m.pubkey)[:4]
+        return fingerprint
+
+    def get_pubkey_at_path(self, bip32_path: str) -> ExtendedKey:
+        """
+        Get the public key at the BIP 32 derivation path.
+
+        :param bip32_path: The BIP 32 derivation path
+        :return: The extended public key
+        """
+        self.interactive_get_cvc()
+        int_path = str2path(bip32_path)
+        check_bip32_path(int_path)  # raises ValueError if hardened after non-hardened
+        hardened, non_hardened = split_bip32_path(int_path)
+        if not non_hardened:
+            self.device.set_derivation(path=bip32_path, cvc=self.password)
+            xp = self.device.get_xpub(self.password)
+            return ExtendedKey.deserialize(xp)
+        else:
+            str_path = path2str(hardened)
+            self.device.set_derivation(path=str_path, cvc=self.password)
+            xp = self.device.get_xpub(self.password)
+            ext_key = ExtendedKey.deserialize(xp)
+            return ext_key.derive_pub_path(non_hardened)
```

Comment:
Does the Tapsigner only do hardened derivation? I don't generally like to have HWI do path validation and would much rather the device do it and let it give us an error.

Reply:
yes, tapsigner only does hardened derivation... I moved the code to `coinkite-tap-protocol` https://github.com/coinkite/coinkite-tap-proto/pull/24 - after it is merged I will update `pyproject.toml` and `setup.py` with current version (for now there is a workaround)